File: longest-subarray-of-1s-after-deleting-one-element/longest-subarray-of-1s-after-deleting-one-element.cpp
class Solution {
public:
    int longestSubarray(vector<int>& nums) {
        int n = nums.size(), sol = -1;
        vector<int> pre (n, 0), post(n, 0);

        pre[0] = nums[0];
        for (int i = 1; i < n; i++) {
            pre[i] = nums[i] == 0 ? 0 : pre[i - 1] + 1;
        } //for

        post[n - 1] = nums[n - 1];
        for (int i = n - 2; i >= 0; i--) {
            post[i] = nums[i] == 0 ? 0 : post[i + 1] + 1;
        } //for

        for (int i = 0; i < n; i++) {
            if (nums[i] == 0) {
                if (i == 0) {
                    sol = max(sol, post[1]);
                } else if (i == n - 1) {
                    sol = max(sol, pre[n - 2]);
                } else {
                    sol = max(sol, pre[i - 1] + post[i + 1]);
                } //if
            } //if
        } //for

        return sol == -1 ? n - 1 : sol;
    } //longestSubarr
};

File: LeetCode/1586-longest-subarray-of-1s-after-deleting-one-element/longest-subarray-of-1s-after-deleting-one-element.java
class Solution {
    public int longestSubarray(int[] nums) {
        int l = 0, r = 0, n = nums.length, 
            used = 0, sol = 0;

        while (r < n) {
            if (nums[r] == 0 && used == 1) {
                if (r - l > sol) sol = r - l;
                while (used == 1) {
                    if (nums[l++] == 0) used--;
                } //while
            } else if (nums[r++] == 0) used = 1;
        } //while

        return Math.max(r - l, sol) - 1;
    } //longestSubArr
} //Sol